plugins/nf-tower/src/main/io/seqera/tower/plugin/TowerFusionEnv.groovy

```diff
@@ -0,0 +1,257 @@
+
+        final retryOnException = (e -> e instanceof IOException) as Predicate<? extends Throwable>
+        final retryOnStatusCode = ((HttpResponse<T> resp) -> resp.statusCode() in errorsToRetry) as Predicate<HttpResponse<T>>
+
+        final listener = new EventListener<ExecutionAttemptedEvent<HttpResponse<T>>>() {
+            @Override
+            void accept(ExecutionAttemptedEvent event) throws Throwable {
+                def msg = "connection failure - attempt: ${event.attemptCount}"
+                if (event.lastResult != null)
+                    msg += "; response: ${event.lastResult}"
+                if (event.lastFailure != null)
+                    msg += "; exception: [${event.lastFailure.class.name}] ${event.lastFailure.message}"
+                log.debug(msg)
+            }
+        }
+        return RetryPolicy.<HttpResponse<T>> builder()
+            .handleIf(retryOnException)
+            .handleResultIf(retryOnStatusCode)
+            .withBackoff(DEFAULT_RETRY_POLICY_DELAY.toMillis(), DEFAULT_RETRY_POLICY_MAX_DELAY.toMillis(), ChronoUnit.MILLIS)
+            .withMaxAttempts(DEFAULT_RETRY_POLICY_MAX_ATTEMPTS)
+            .withJitter(DEFAULT_RETRY_POLICY_JITTER)
+            .onRetry(listener)
+            .build()
+    }
+
+    /**
+     * Send an HTTP request and return the response. This method automatically retries the request according to the
+     * given RetryPolicy.
+     *
+     * @param req The HttpRequest to send
+     * @return The HttpResponse received
+     */
+    private <T> HttpResponse<String> safeHttpSend(HttpRequest req, RetryPolicy<T> policy) {
+        return Failsafe.with(policy).get(
+            () -> {
+                log.debug "Request: method:=${req.method()}; uri:=${req.uri()}; request:=${req}"
+                final resp = httpClient.send(req, HttpResponse.BodyHandlers.ofString())
+                log.debug "Response: statusCode:=${resp.statusCode()}; body:=${resp.body()}"
+                return resp
+            } as CheckedSupplier
+        ) as HttpResponse<String>
+    }
+
+    /**
+     * Create a JSON string representing a {@link LicenseTokenRequest} object
+     *
+     * @param product The product SKU
+     * @param version The version
+     * @return The resulting JSON string
+     */
+    private static String makeLicenseTokenRequest(String product, String version) {
+        return new Gson().toJson(
+            new LicenseTokenRequest(
+                product: product,
+                version: version
+            ),
+            LicenseTokenRequest.class
+        )
+    }
+
+    /**
+     * Parse a JSON string into a {@link LicenseTokenResponse} object
+     *
+     * @param stringHttpResponse The HttpResponse containing the JSON string
+     * @return The resulting LicenseTokenResponse object
+     */
+    private static LicenseTokenResponse parseLicenseTokenResponse(HttpResponse<String> resp) {
```

Comment:
Does this need to be static? It's private and it seems to be called only by non static method `getLicenseToken`

Reply:
Thanks consider this resolved, I apparently do not have the ability to resolve myself :(